SimTKcommon/Simulation/include/SimTKcommon/internal/StateImpl.h

```diff
@@ -126,130 +234,115 @@ class DiscreteVarInfo {
 //==============================================================================
 //                            CACHE ENTRY INFO
 //==============================================================================
-class CacheEntryInfo {
+/* A cache entry holds an AbstractValue that is computed ("realized") from the
+values of state variables. Its primary purpose is to efficiently and flawlessly
+keep track of whether that value is up to date with respect to the state
+variables from which it is realized. For efficiency, there are two 
+mechanisms used: the "computation stage" system provides coarse granularity
+that is sufficient for most cache entries. When additional finesse is needed,
+a cache entry may also specify a set of prerequisite state variables or other
+cache entries that it depends on. 
+
+A cache entry specifies a "depends-on" stage. If its owning subsystem's stage
+is below that, the cache entry's value *cannot* be valid. It may optionally
+specify a "computed-by" stage; if its subsystem's stage has reached that then
+the cache entry's value is *guaranteed* to be valid. If the stage is in 
+between, then validity is determined as follows:
+  - compare the saved depends-on stage version with the curent version; if they
+    don't match then the cache entry is not valid
+  - then, the cache entry is valid if its up-to-date-with-prerequisites
+    flag is set and invalid otherwise.
+
+Cache entries also have an allocation stage (Topology, Model, or Instance)
+and must be de-allocated if that stage is invalidated.
+*/
+class SimTK_SimTKCOMMON_EXPORT CacheEntryInfo {
```

Comment:
> Why export? Isn't this an internal type? DiscreteVarInfo is not exported.

It's because of several routines here that are not inlined, like `registerWithPrerequisites()`. Those are in the DLL. DiscreteVarInfo is entirely inline. I could move the export to each of the non-inlined methods but it seems easier this way. I've currently got worse problems with the force-inline stuff on gcc & clang (different problems on each).